debug.go

```diff
@@ -0,0 +1,66 @@
+package basictracer
+
+import (
+	"bytes"
+	"fmt"
+	"runtime"
+	"strconv"
+	"sync"
+)
+
+const debugGoroutineIDTag = "_initial_goroutine"
+
+type errAssertionFailed struct {
+	msg string
+}
+
+func (s *spanImpl) Lock() {
+	s.Mutex.Lock()
+	s.maybeAssertLocked()
+}
+
+func (s *spanImpl) maybeAssertLocked() {
+	if s.tracer.Options.DebugAssertSingleGoroutine {
+		startID := curGoroutineID()
+		curID, ok := s.raw.Tags[debugGoroutineIDTag].(uint64)
+		if !ok {
+			// This is likely invoked in the context of the SetTag which sets
+			// debugGoroutineTag.
+			return
+		}
+		if startID != curID {
+			panic(&errAssertionFailed{
+				msg: fmt.Sprintf("span started on goroutine %d, but now running on %d", startID, curID),
+			})
+		}
+	}
+}
+
+var goroutineSpace = []byte("goroutine ")
+var littleBuf = sync.Pool{
+	New: func() interface{} {
+		buf := make([]byte, 64)
+		return &buf
+	},
+}
+
+// Credit to @bradfitz:
```

Comment:
Have you seen the version that generates random combinations of nested function calls in conjunction with the PCs in the stack frames to identify particular goroutines??? really epic! From the readme:

> ## What are people saying?
> 
> "Wow, that's horrifying."
> 
> "This is the most terrible thing I have seen in a very long time."

https://github.com/jtolds/gls/blob/master/context.go#L130